File: app/src/main/java/alejandromr92/com/movies/presentation/ui/adapters/MovieListAdapter.java
package alejandromr92.com.movies.presentation.ui.adapters;

import alejandromr92.com.movies.R;
import alejandromr92.com.movies.presentation.model.MovieView;
import alejandromr92.com.movies.presentation.ui.adapters.holders.MovieHolder;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class MovieListAdapter extends RecyclerView.Adapter<MovieHolder> implements Filterable {

    private List<MovieView> items;
    private List<MovieView> filteredItems;
    private WeakReference<Context> reference;

    public MovieListAdapter(List<MovieView> items, WeakReference<Context> reference) {
        this.items = items;
        this.filteredItems = items;
        this.reference = reference;
    }

    @NonNull
    @Override
    public MovieHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.item_movie, parent, false);

        return new MovieHolder(view, reference);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieHolder holder, int position) {
        holder.bind(filteredItems.get(position));
    }

    @Override
    public int getItemCount() {
        return filteredItems != null ? filteredItems.size() : 0;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                String charString = charSequence.toString();
                if (charString.isEmpty()) {
                    filteredItems = items;
                } else {
                    List<MovieView> moviesFilteredList = new ArrayList<>();
                    for (MovieView row : items) {
                        if (row.getTitle().toLowerCase().contains(charString.toLowerCase())){
                            moviesFilteredList.add(row);
                        }
                    }

                    filteredItems = moviesFilteredList;
                }

                FilterResults filterResults = new FilterResults();
                filterResults.values = filteredItems;
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                filteredItems = (List<MovieView>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }
}

File: app/src/main/java/alejandromr92/com/movies/presentation/ui/adapters/MovieListAdapter.kt
package alejandromr92.com.movies.presentation.ui.adapters

import alejandromr92.com.movies.R
import alejandromr92.com.movies.presentation.model.MovieView
import alejandromr92.com.movies.presentation.ui.adapters.holders.MovieHolder
import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable

import java.lang.ref.WeakReference
import java.util.ArrayList

class MovieListAdapter(private val items: List<MovieView>, private val reference: WeakReference<Context>) :
    RecyclerView.Adapter<MovieHolder>(), Filterable {
    private var filteredItems: List<MovieView>? = null

    init {
        this.filteredItems = items
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieHolder {
        val view = LayoutInflater
            .from(parent.context)
            .inflate(R.layout.item_movie, parent, false)

        return MovieHolder(view)
    }

    override fun onBindViewHolder(holder: MovieHolder, position: Int) {
        holder.bind(filteredItems!![position])
    }

    override fun getItemCount(): Int {
        return if (filteredItems != null) filteredItems!!.size else 0
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): Filter.FilterResults {
                val charString = charSequence.toString()
                if (charString.isEmpty()) {
                    filteredItems = items
                } else {
                    val moviesFilteredList = ArrayList<MovieView>()
                    for (row in items) {
                        if (row.title.toLowerCase().contains(charString.toLowerCase())) {
                            moviesFilteredList.add(row)
                        }
                    }

                    filteredItems = moviesFilteredList
                }

                val filterResults = Filter.FilterResults()
                filterResults.values = filteredItems
                return filterResults
            }

            override fun publishResults(charSequence: CharSequence, filterResults: Filter.FilterResults) {
                filteredItems = filterResults.values as List<MovieView>
                notifyDataSetChanged()
            }
        }
    }
}
